pcs/utils.py

```diff
@@ -281,3 +291,38 @@ def copy_find_and_replace_leaves(
         else:
             transform_leaf(data_struct, copy_struct, leaf_list, lambda x: x)
     return match_found, copy_struct
+
+
+def pad_list_if_necessary(list_in, index):
+    while index >= len(list_in):  # List too short
+        list_in.append(None)
+
+
+def pipe_and_check_popen(args: List[str], **kwargs):
```

Comment:
Nice!  Looks like you're using this to call out to pip.  Seems like a smarter (or at least more customizable) version of `subprocess.run`.  What prompted you to go about it this way?

Reply:
I found myself tweaking the flags we pass to Popen so that we fail fast when our sub processes fail (e.g., pip install into a virtualenv) and the stdout/stderr show up to make debugging easier. And I figured DRYing out the various places where we use subprocesses (to run the `virtualenv` command vs to run the `pip` command) would be good hygiene.